Move docker image management and test entrypoint to Maven

mccheah opened the pull request:
Closes #30.

mccheah committed:
Use an nginx server for remote jars tests.

mccheah committed:
Moves all integration test setup logic to Maven and scripts.

The Kubernetes integration tests now always expect an image to be
pre-built, so we no longer build images with Scala code. Maven's
pre-integration-test invokes a single script to bootstrap the
environment with the built images, etc. In the transition we try to keep
as much of the same semantics as possible.

mccheah committed:
Update documentation

mccheah commented:
Requires both #29 and #14.

mccheah commented:
Also requires the test commands in Jenkins to change.

mccheah committed:
Remove unnecessary .gitignore entries

mccheah commented:
todo and not handled by this patch:

- Clean up docker images in the `post-integration-test` phase
- Use your own docker images (it's theoretically possible but the usage of the examples file server and the usage of the same tag between the file server and the Spark code under test makes this workflow less than intuitive)

mccheah commented on build/mvn:
```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env bash
```
Taken from https://github.com/apache/spark/blob/master/build/mvn

foxish replied:
Why do we need this? We'll need to ensure that it stays in sync with our repo? Or maybe we can get rid of it once we merge this upstream?

mccheah replied:
We get rid of this when we merge into upstream, but even aside from that, this has hardly been changing in upstream and it's theoretically completely isolated from what we will use to build Spark (since the shell forks the second Maven process it's completely independent)

mccheah replied:
We also have to add it here now because unlike before, now `mvn` is the entrypoint into the whole system, as it's only the substep of the build reactor that clones the Spark repository down. Aside from that, we also allow for Spark TGZs to be provided to remove the usage of the Spark source code entirely. So no matter what we're going to need our own Maven here.

ssuchter replied:
Instead of copying the content here, could you just have a much shorter script that just downloads from https://raw.githubusercontent.com/apache/spark/master/build/mvn and execs it? Something like build/mvn-getter whose content is:
```
#!/bin/sh
curl -s https://raw.githubusercontent.com/apache/spark/master/build/mvn | bash

mccheah replied:
Hm, I think that works. Will incorporate that.

mccheah replied:
Done

mccheah commented on scripts/parse-arguments.sh:
```diff
@@ -0,0 +1,82 @@
+#
+
+set -ex
+TEST_ROOT_DIR=$(git rev-parse --show-toplevel)
```
A lot of these defaults are repeated again in `pom.xml` - the idea is that maintainers of the framework can run the scripts in isolation without using Maven. This can be done both with the top level `setup-integration-test-env.sh` script, or any of the sub-components - e.g. `clone-spark.sh`.

foxish replied:
Hmm.. it seems error prone that this needs to stay in sync with the args in pom.xml. Is there a better dispatch mechanism than to use this?

mccheah replied:
The `pom.xml` can assume that these are the defaults and that they are correct, but I like having them explicitly enumerated in the `pom.xml` because the behavior can be known right from the pom. But I also like being able to run the script independently without any arguments. But all of this is just for convenience, so for the sake of developer sanity we can make one of these ideas less convenient.

Don't have a strong opinion on the best way to move forward.

mccheah commented on scripts/parse-arguments.sh:
```diff
@@ -0,0 +1,82 @@
+DEPLOY_MODE=minikube
+IMAGE_REPO="docker.io/kubespark"
+SKIP_BUILDING_IMAGES=false
+SPARK_TGZ="N/A"
```
This is less than ideal. It's more elegant to have here: `SPARK_TGZ=` and then to check in other places, `if [[ -z $SPARK_TGZ ]]`. The problem is that Maven's `exec` plugin has to pass in a fixed set of command line arguments. Maven will always have to pass in some `spark-tgz` argument. But we can't pass in empty, because we then end up with commands like `setup-integration-test-env --spark-tgz --deploy-mode cloud` which screws up the argument shifting.

An alternative we can employ in general is to allow these variables to be set by environment variables as well and to pass them in as environment variables from Maven. Ideas are welcome here.

ssuchter replied:
You could use the maven exec plugin like this:

```
              <executable>/bin/sh</executable>
              <arguments>
                <argument>-c</argument>
                <argument>arbitrary length here ; another command</argument>
              </arguments>
```

mccheah replied:
The problem is that the arguments themselves have to be fixed. I can't sometimes pass in the `--spark-tgz` flag and sometimes not pass the `--spark-tgz` flag.

ssuchter replied:
This gets simpler with the idea that this code always takes --spark-tgz, I think.

mccheah committed:
Use $IMAGE_TAG instead of $TAG

mccheah committed:
Don't write image tag file twice

mccheah commented on scripts/write-docker-tag.sh:
```diff
@@ -0,0 +1,30 @@
+fi
+
+rm -f $IMAGE_TAG_OUTPUT_FILE
+touch $IMAGE_TAG_OUTPUT_FILE
```
Less than ideal. The problem is that Maven wants to pass the image tag into `KubernetesSuite`, but Maven doesn't know it offhand until it's resolved by the test environment script. We use a UUID for the default image tag because there can be multiple builds concurrently running against the same Spark git hash. But the UUID can't be generated by Maven, meaning the script has to make it and pass that information along somehow, and a temporary file was the best bridge I could think of. Other more elegant ideas would be appreciated here.

foxish replied:
Concurrent runs against the same git-sha, I'm not quite sure. So, if a new PR is submitted, it would apply that commit on the base and then run the tests - which would make each one have a unique sha. If they are the same sha, then invariably, they are identical and the image could be reused.

mccheah replied:
Think there's two git hashes to consider and neither will work:

- If we're talking about the integration test repository git hash, multiple builds should certainly be able to run against the same commit. Multiple test runs on the same node testing two different commits of Spark but using the same integration test version comes to mind.
- If we're talking about the Spark repository git hash, the inverse rule applies here - multiple versions of the integration test running against the same Spark version. Reusing the images is difficult because we need to know when it's safe to GC the image, which is only if all tests holding that given tag are finished. Then we can run into race conditions, etc.

Seems a lot easier to bias towards isolation for correctness and optimize later.

ssuchter replied:
Would it be simpler to just GC any images that are from this, but over a day old?

mccheah replied:
There's still a potential race with who gets to build the image with the given hash though. Two builds trying to use the same hash need to agree on who builds the image. I don't want to have to deal with that locking at all, so just use completely unique tags everywhere.

ssuchter replied:
Ok!

mccheah commented:
The diff is against a branch that is the outcome of merging master with #29 and #14.

foxish commented on e2e/e2e-prow.sh:
```diff
@@ -1,39 +0,0 @@
-#!/bin/bash
```
This is specific to GKE testsuites, I think we still need this setup script, even if the runner goes away.

mccheah replied:
Ah I didn't notice the apt-get calls. I think we want these to be assumed by the framework. Installing components on demand in a test framework is scary-ish - not to mention that installation probably depends on the user's OS, e.g. if the user is running from Mac OSX this won't work. We should have separate scripts that aren't invoked in the build lifecycle for these.

foxish replied:
Yes, did not intend for these to be part of the testsuite. These are external "setup" steps and the entrypoint for the k8s CI system.

foxish commented:
Agreed. It's separate and the only expected consumer is k8s test
infrastructure.

On Jan 12, 2018 2:42 PM, "mccheah" <notifications@github.com> wrote:

> *@mccheah* commented on this pull request.
> ------------------------------
>
> In e2e/e2e-prow.sh
> <https://github.com/apache-spark-on-k8s/spark-integration/pull/31#discussion_r161342654>
> :
>
> > @@ -1,39 +0,0 @@
> -#!/bin/bash
>
> Ah I didn't notice the apt-get calls. I think we want these to be assumed
> by the framework. Installing components on demand in a test framework is
> scary-ish - not to mention that installation probably depends on the user's
> OS, e.g. if the user is running from Mac OSX this won't work. We should
> have separate scripts that aren't invoked in the build lifecycle for these.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache-spark-on-k8s/spark-integration/pull/31#discussion_r161342654>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3U5zFSnkZLVZef7SZqoSf2Xwg__39uks5tJ9_lgaJpZM4Rc4bM>
> .
>

foxish commented on README.md:
```diff
@@ -8,98 +8,71 @@ title: Spark on Kubernetes Integration Tests
+## Re-using Docker Images
+
+By default, the test framework will build new Docker images on every test execution. A unique image tag is generated,
+and it is written to file at `target/imageTag.txt`. To reuse the images built in a previous run, set:
```
I liked before that it was using the git sha - it made it easy to correlate to the distribution that was just tested.

mccheah replied:
Discussed in https://github.com/apache-spark-on-k8s/spark-integration/pull/31#discussion_r161349039

foxish commented on README.md:
```diff
@@ -8,98 +8,71 @@ title: Spark on Kubernetes Integration Tests
+By default, the test framework will test the master branch of Spark from [here](https://github.com/apache/spark). You
+can specify the following options to test against different source versions of Spark:
+
+* `spark.kubernetes.test.sparkRepo` to the git or http URI of the Spark git repository to clone
```
`gitRepo` and `gitBranch` may be better names for this.

foxish commented on scripts/build-spark.sh:
```diff
@@ -0,0 +1,33 @@
+mkdir -p $UNPACKED_SPARK_TGZ
+if [[ $SPARK_TGZ == "N/A" ]];
+then
+  ./dev/make-distribution.sh --tgz -Phadoop-2.7 -Pkubernetes -DskipTests;
```
If we omit `--tgz`, we should just get the unpacked distribution under `dist/`.

mccheah replied:
Makes it easier to reuse the distribution with the `spark.kubernetes.test.sparkTgz` option. Thus that configuration can pull double duty with custom Spark distributions and re-using dynamically built ones.

ssuchter replied:
I think for jenkins, we will want to use the pre-built distributions exclusively. This is because we don't want to keep the minikube locked during the building of the distribution. If the spark distribution builds "inside" of the spark-integration test, then it will be harder to express this partial locking. As it currently stands, one jenkins job creates the spark distribution, and another jenkins job does the integration test, making the locking easy.

mccheah committed:
Remove using nginx file server

foxish commented on src/test/scala/org/apache/spark/deploy/k8s/integrationtest/config.scala:
```diff
@@ -0,0 +1,43 @@
+
+package object config {
+  val KUBERNETES_TEST_DOCKER_TAG_SYSTEM_PROPERTY = "spark.kubernetes.test.imageDockerTag"
+  val DRIVER_DOCKER_IMAGE = "spark.kubernetes.driver.container.image"
```
These need to be cleaned up I think.

mccheah commented:
Changed the base to `master` and removed dependence on #29. Turns out we actually need this to run at all now - our tests are broken since the Docker image format change introduced in https://github.com/apache/spark/pull/20192

mccheah committed:
Remove some lines

foxish commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,34 @@
+
+if [[ $SPARK_TGZ == "N/A" ]];
+then
+  $SCRIPTS_DIR/clone-spark.sh "$@";
```
Instead of passing all the args around, we should parge the args here, and then set the right env-vars/args and call the other scripts maybe? I think it's an antipattern for each script to be calling parse-args and seeing unrelated args.

ssuchter commented on README.md:
```diff
@@ -8,98 +8,71 @@ title: Spark on Kubernetes Integration Tests
+      -Dspark.kubernetes.test.sparkRepo=https://github.com/apache-spark-on-k8s/spark \
+      -Dspark.kubernetes.test.sparkBranch=new-feature
+
+Additionally, you can use a pre-built Spark distribution. In this case, the repository is not cloned at all, and no
```
I don't understand the motivation of why we should have the integration test code ever try to clone and build spark. Why not just always depend on a pre-built spark distribution? (Sorry if I missed the reason for this)

mccheah replied:
I've found it a lot easier locally to be able to run a single command that both fetches the Spark distribution and builds it and then the integration test uses that Spark distribution. We provide the optionality for the local development scenario but we can still provide the TGZ specifically in Jenkins via `spark.kubernetes.test.sparkTgz`.

ssuchter commented on scripts/parse-arguments.sh:
```diff
@@ -0,0 +1,82 @@
+
+set -ex
+TEST_ROOT_DIR=$(git rev-parse --show-toplevel)
+BRANCH="master"
```
Presumably if we don't try to build spark in here, we don't need BRANCH, SPARK_REPO, SPARK_REPO_LOCAL_DIR, UNPACKED_SPARK_TGZ (we might only have SPARK_TGZ, and it's always required), right?

ssuchter commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,34 @@
+SCRIPTS_DIR=$(dirname $0)
+source $SCRIPTS_DIR/parse-arguments.sh "$@"
+
+if [[ $SPARK_TGZ == "N/A" ]];
```
I think if we have this always take a spark.tgz, then this branch isn't necessary

ssuchter commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,34 @@
+then
+  $SCRIPTS_DIR/clone-spark.sh "$@";
+fi
+$SCRIPTS_DIR/build-spark.sh "$@"
```
Maybe this would change to unpack-spark.sh?

ssuchter commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,34 @@
+fi
+$SCRIPTS_DIR/build-spark.sh "$@"
+
+if [[ $SKIP_BUILDING_IMAGES == false ]] ;
```
If unpacking spark is something we always do, is there any reason to have this skip argument? Why else would one call this setup-integration-test-env.sh other than building the dockers?

mccheah committed:
Split building Spark for dev environment from build reactor

mccheah commented:
@ssuchter @foxish approach has changed - the `dev/dev-run-integration-tests.sh` is for local running and is whats documented in `README.md`, which builds the Spark distribution and sends it to the Maven reactor. The maven reactor itself uses a separate script to build the Docker image, but the Maven reactor must be given the Spark distribution to run with.

mccheah commented:
Reason we have to do this was discussed in sig-big-data - Jenkins will only use the Maven build without building the Spark distribution, so that the maven build doesn't lock the Minikube resource on building Spark, which takes a long time.

mccheah committed:
Small docs fix

mccheah committed:
Docs formatting fix

mccheah committed:
Spark TGZ can be empty instead of N/A, throw an error if not provided.

mccheah committed:
Remove extraneous --skip-building-docker-images flag.

mccheah committed:
Switch back to using the N/A placeholder

mccheah committed:
Remove extraneous code

mccheah commented on src/test/scala/org/apache/spark/deploy/k8s/integrationtest/backend/GCE/GCETestBackend.scala:
```diff
@@ -36,8 +36,4 @@ private[spark] class GCETestBackend(val master: String) extends IntegrationTestB
   override def getKubernetesClient(): DefaultKubernetesClient = {
     defaultClient
   }
-
```
Though the more I look at this abstraction, the more I realize it's not entirely necessary anymore. The biggest difference before was that using the MinikubeTestBackend triggered the Docker build, but now, we're building images regardless of the backend type. The Minikube version accepts the Minikube certificates, but those can be passed along as system properties. It's not something to change in this PR, but it's something that can be changed strictly because of the way the image build workflow functions now.

mccheah commented:
@foxish @ssuchter the command in Jenkins is now incorrect. Seeing that all our other builds will fail now anyways, we might as well update the build command to reflect what this version of the tests expects. Then we can check if this build would be green.

The Jenkins instance ran this:

`/home/jenkins/tools/hudson.tasks.Maven_MavenInstallation/maven/bin/mvn -f integration-test/pom.xml clean integration-test -DzincPort=3096 -Dspark-distro-tgz=/home/jenkins/workspace/PR-dev-spark-integration/target/copied-artifacts-spark-distro/spark-2.4.0-SNAPSHOT-bin-20180112-55dbfbc.tgz`

Pretty sure it has to be changed to this now:

    <repo-root>/build/mvn clean integration-test -Dspark.kubernetes.test.sparkTgz=<spark tgz path> -DzincPort=3096

Main differences of course being, we now use our built-in Maven, and we can use the `pom.xml` file verbatim. But this PR also changes the property for the Spark distribution, mostly just for aesthetics.

mccheah committed:
Remove maven args because they don't work

mccheah commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+if [ -z $SPARK_MASTER ];
+then
+  build/mvn integration-test \
+    -Dspark.kubernetes.test.sparkTgz=$SPARK_TGZ \
```
Would like to add support for arbitrary maven arguments, just couldn't quite get the shell scripting to work out yet.

ssuchter commented on build/mvn:
```diff
@@ -14,23 +15,10 @@
-./e2e/runner.sh -m $master -r $repo -i $image_repo -d minikube
+BUILD_DIR=$(dirname $0)
+MVN_RUNNER=$BUILD_DIR/run-mvn
+curl -s https://raw.githubusercontent.com/apache/spark/master/build/mvn > $MVN_RUNNER
```
Why not just pipe to bash?

mccheah replied:
Do the arguments sent to this script follow the pipe?

mccheah replied:
We also don't want to download the mvn script if we already have it - so the script ends up writing simpler with that fix. Addressed in latest patch.

ssuchter commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+  shift
+done
+
+if [[ $SPARK_TGZ == "N/A" ]];
```
Are you going to keep this clause, or delete it? I think it's pretty straightforward this way, so I think it'd be reasonable to keep it.

ssuchter commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+#
+
+TEST_ROOT_DIR=$(git rev-parse --show-toplevel)
+BRANCH="master"
```
Remove unnecessary quotes

ssuchter commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+
+TEST_ROOT_DIR=$(git rev-parse --show-toplevel)
+BRANCH="master"
+SPARK_REPO="https://github.com/apache/spark"
```
Remove unnecessary quotes

ssuchter commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+SPARK_REPO="https://github.com/apache/spark"
+SPARK_REPO_LOCAL_DIR="$TEST_ROOT_DIR/target/spark"
+DEPLOY_MODE=minikube
+IMAGE_REPO="docker.io/kubespark"
```
Remove unnecessary quotes

ssuchter commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+SPARK_REPO_LOCAL_DIR="$TEST_ROOT_DIR/target/spark"
+DEPLOY_MODE=minikube
+IMAGE_REPO="docker.io/kubespark"
+SPARK_TGZ="N/A"
```
Remove unnecessary quotes

ssuchter commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+DEPLOY_MODE=minikube
+IMAGE_REPO="docker.io/kubespark"
+SPARK_TGZ="N/A"
+IMAGE_TAG="N/A"
```
Remove unnecessary quotes

ssuchter commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,92 @@
+UNPACKED_SPARK_TGZ="$TEST_ROOT_DIR/target/spark-dist-unpacked"
+IMAGE_TAG_OUTPUT_FILE="$TEST_ROOT_DIR/target/image-tag.txt"
+DEPLOY_MODE=minikube
+IMAGE_REPO="docker.io/kubespark"
```
Remove unnecessary quotes

ssuchter commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,92 @@
+IMAGE_TAG_OUTPUT_FILE="$TEST_ROOT_DIR/target/image-tag.txt"
+DEPLOY_MODE=minikube
+IMAGE_REPO="docker.io/kubespark"
+IMAGE_TAG="N/A"
```
Remove unnecessary quotes

ssuchter commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,92 @@
+DEPLOY_MODE=minikube
+IMAGE_REPO="docker.io/kubespark"
+IMAGE_TAG="N/A"
+SPARK_TGZ="N/A"
```
Remove unnecessary quotes

ssuchter commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,92 @@
+
+if [[ $SPARK_TGZ == "N/A" ]];
+then
+  echo "Must specify a Spark tarball to build Docker images against with --spark-tgz." && exit 1;
```
Remove unnecessary quotes

ssuchter reviewed (approved):
I think this looks nice. I mainly left some comments about removing some unneeded double quotes in shell expressions. :)

ssuchter commented:
I should say to all of my comments about removing unnecessary quotes. I guess it'd also be reasonable to add quotes for all variable assignments (although they aren't necessary). I think consistency is important.

kimoonkim commented:
@mccheah wrote
> @foxish @ssuchter the command in Jenkins is now incorrect. Seeing that all our other builds will fail now anyways, we might as well update the build command to reflect what this version of the tests expects. Then we can check if this build would be green.

This sounds good to me. Has anyone started fixing this? Otherwise, I'll take on that. Builds are all failing as predicted and we'd better fix it soon.

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

foxish commented:
@kimoonkim, I don't think anyone has looked at the jenkins command yet. Would be good if we can fix it in time for this PR to merge and get things green again.

foxish commented on scripts/setup-integration-test-env.sh:
```diff
@@ -0,0 +1,92 @@
+    fi
+  else
+    # -m option for minikube.
+    eval $(minikube docker-env)
```
This would already be done by the `-m` option.

kimoonkim commented:
Cool. Working on it.

kimoonkim commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+    -Dspark.kubernetes.test.sparkTgz=$SPARK_TGZ \
+    -Dspark.kubernetes.test.imageTag=$IMAGE_TAG \
+    -Dspark.kubernetes.test.imageRepo=$IMAGE_REPO \
+    -Dspark.kubernetes.test.deployMode=$DEPLOY_MODE \
```
Should drop the backslash here.

ssuchter replied:
Duh, it's just the last line (line 92) where it should be dropped, I got it now.

ssuchter replied:
I think this is addressed.

kimoonkim commented on dev/dev-run-integration-tests.sh:
```diff
@@ -0,0 +1,100 @@
+    -Dspark.kubernetes.test.imageTag=$IMAGE_TAG \
+    -Dspark.kubernetes.test.imageRepo=$IMAGE_REPO \
+    -Dspark.kubernetes.test.deployMode=$DEPLOY_MODE \
+    -Dspark.kubernetes.test.master=$SPARK_MASTER \
```
Should drop the backslash.

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
Ok. I made Jenkins use dev-run-integration-tests.sh, which get us the following error:
> dev/dev-run-integration-tests.sh --spark-tgz /home/jenkins/workspace/PR-dev-spark-integration/target/copied-artifacts-spark-distro/spark-2.4.0-SNAPSHOT-bin-20180113-fc6fe8a.tgz
dev/dev-run-integration-tests.sh: line 101: syntax error: unexpected end of file

I think it's due to the bug pointed out [here](https://github.com/apache-spark-on-k8s/spark-integration/pull/31#pullrequestreview-88665624)

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

kimoonkim commented:
rerun integration tests please

mccheah commented:
@kimoonkim the `dev/dev-run-integration-tests.sh` script is for development setups, though I took a look at the command and that command should theoretically work in Jenkins as well.

Couple of questions:

- Minikube appears to be starting in the output, shouldn't it already be running?
- Minikube is stopping at the end. Shouldn't we be keeping it running?

mccheah committed:
Fix scripts

mccheah committed:
Don't get Maven if it's already there

mccheah commented:
@ssuchter let's just have quotes everywhere.

mccheah commented:
@kimoonkim the `minikube` binary needs to be on the `PATH`.

mccheah committed:
Put quotes everywhere

mccheah commented:
Sorry, forgot to mention what the specific command should be in Jenkins without using the dev script: `build/mvn integration-test -Dspark.kubernetes.test.sparkTgz=<tgz-path>`.

mccheah commented:
rerun integration tests please

mccheah commented:
@kimoonkim integration tests aren't running when I comment, can you trigger the build?

kimoonkim commented:
> @kimoonkim integration tests aren't running when I comment, can you trigger the build?

I think this build [98](http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-dev-spark-integration/98/) actually ran upon your comment, given the timing of build start. I believe there is a quiet period of dozen secs. So maybe it did not trigger right away.

mccheah committed:
Merge remote-tracking branch 'origin/master' into move-all-logic-to-maven

ssuchter reviewed (commented):
.

mccheah committed:
Minor formatting

kimoonkim commented:
> Minikube appears to be starting in the output, shouldn't it already be running?
Minikube is stopping at the end. Shouldn't we be keeping it running?

Right. It gets stopped and deleted at the end. Pepperdata Jenkins cluster has only one node and it still has jobs running old integration tests from our fork. If we keep minikube running, that will interfere with the old integration tests. Stopping minikube is the best compromise for this setup, IMO.

mccheah commented:
Hm, but the error is this:

```
  java.io.IOException: Cannot run program "minikube": error=2, No such file or directory
  at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
```

But the tests work locally for me. I wonder if Java's `ProcessBuilder` doesn't use the `PATH` to find locations of executables.

kimoonkim commented:
> Sorry, forgot to mention what the specific command should be in Jenkins without using the dev script: build/mvn integration-test -Dspark.kubernetes.test.sparkTgz=<tgz-path>.

Jenkins can switch to this command, but what is the advised way of building docker images now that the scala code does not build them?

mccheah commented:
@kimoonkim this PR puts us in an in-between state where Maven is building the Docker images but we expect the external Jenkins job to build the Spark distribution. If we build using Maven, the reactor will still build the images for us.

kimoonkim commented:
> Hm, but the error is this:
>>
>>  java.io.IOException: Cannot run program "minikube": error=2, No such file or directory
  at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
> But the tests work locally for me. I wonder if Java's ProcessBuilder doesn't use the PATH to find locations of executables.

I see I'll try adding the minikube location to the PATH used by maven.

kimoonkim commented:
rerun integration tests please

mccheah committed:
Hard set Minikube binary location.

kimoonkim commented:
Now, the error from build [101](http://spark-k8s-jenkins.pepperdata.org:8080/job/PR-dev-spark-integration/101/consoleFull) is this:
> /home/jenkins/workspace/PR-dev-spark-integration/target/spark-dist-unpacked/bin/docker-image-tool.sh: line 66: docker: command not found

@ssuchter I believe you installed minikube on Pepperdata Jenkins. Did we install docker?

ssuchter commented:
We're working on that same issue (missing docker) in slack sig-big-data

mccheah commented:
rerun integration tests please

mccheah commented:
rerun integration tests please

mccheah committed:
Run Minikube from bash -c
